### PR TITLE
Issue #1472.  Displaying all expected and predicted entity mentions in a user specified json file

### PR DESCRIPTION
**Pull request by demongolem**
This pull request implements the feature request in Issue #1472.

**Proposed changes**:
- Added new cmd argument to test.py which specifies a json output file for entity mentions
- Analyze all entity mentions and divide them into true positives, false positives and false negatives to compare the expected and predicted entity mentions

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by akelad** [comment]
Thanks for submitting this PR, we'll give it a review as soon as possible

**Comment by demongolem** [comment]
The checks have passed now and this can be automatically merged, so this PR should be ready for review.  Please take a look at this one

**Comment by erohmensing** [comment]
@demongolem can you move your `test_test` tests into `tests/nlu/base/test_evaluation.py`, where the other testing for that file is?

**Comment by demongolem** [comment]
@erohmensing I have done as requested and the Travis build is on-going, but here is my rationale for naming it as is.

In rasa.nlu.evaluate, I saw the following code line in my copy of the code: `    logger.warning("Calling `rasa.nlu.evaluate` is deprecated. "
                   "Please use `rasa.nlu.test` instead.")`.  So that is why I did not want to use the name evaluate for my test file.

**Comment by demongolem** [comment]
Everything should be ready to go for the review

**Comment by erohmensing** [comment]
@demongolem Great, thank you -- I'll get to this as soon as possible :)

**Comment by akelad** [comment]
@demongolem could you let us know when this is actually ready for review? also, would you be able to squash your commits once you're done please, we can't merge this as it is now

**Comment by demongolem** [comment]
This has been ready for review for weeks now.   From time to time, I have been fetching upsteam to try to keep my fork updated with master so I can work on other things.  There was a build error last night when I did the last fetch last night and it appears to be a docker error not related to my work.

**Comment by akelad** [comment]
yeah i know, and the plan was to review it but then there was a bunch of new commits so we weren't sure whether it was ready anymore. @erohmensing do you have time to take a look at this next week? 
In any case though, we'll need to squash your commits before merging just because there's so many of them

File: rasa/nlu/test.py [hunk 1]
```diff
@@ -442,8 +442,73 @@ def substitute_labels(labels, old, new):
...
+                None if predicted_count == len(pe) else pe[predicted_count]
+            )
+            if predicted_entity is None or (
+                target_entity is not None
+                and target_entity["start"] < predicted_entity["start"]
+            ):
+                fns.append((target_entity, None))
+                target_count += 1
+            elif target_entity is None or (
+                predicted_entity is not None
```
Comment (erohmensing) [inline rasa/nlu/test.py]: we already know `predicted_entity is not None` because we got past line 473

File: rasa/nlu/test.py [hunk 2]
```diff
@@ -442,8 +442,73 @@ def substitute_labels(labels, old, new):
...
+                if (
+                    predicted_type == target_type
+                    and predicted_entity["end"] == target_entity["end"]
+                ):
+                    tps.append((target_entity, predicted_entity))
+                else:
+                    fps.append((target_entity, predicted_entity))
+                    fns.append((target_entity, predicted_entity))
+                target_count += 1
+                predicted_count += 1
```
Comment (erohmensing) [inline rasa/nlu/test.py]: I think in general this logic can be cleaned up by not separating `fps` and `fns` -- I think we can have all errors in one spot -- then you don't have to keep track of whether it's predicted or target that is None and which one's ["end"] comes before the other.

File: rasa/nlu/test.py [hunk 3]
```diff
@@ -442,8 +442,73 @@ def substitute_labels(labels, old, new):
...
+                and predicted_entity["start"] < target_entity["start"]
+            ):
+                fps.append((None, predicted_entity))
+                predicted_count += 1
+            else:
+                target_type = target_entity["entity"]
+                predicted_type = predicted_entity["entity"]
+                if (
+                    predicted_type == target_type
+                    and predicted_entity["end"] == target_entity["end"]
```
Comment (erohmensing) [inline rasa/nlu/test.py]: Let's check that everything is the same, not just entity type and end

File: rasa/nlu/test.py [hunk 4]
```diff
@@ -442,8 +442,73 @@ def substitute_labels(labels, old, new):
...
+                else:
+                    fps.append((target_entity, predicted_entity))
+                    fns.append((target_entity, predicted_entity))
+                target_count += 1
+                predicted_count += 1
+
+    logger.info("Writing entity mentions to file")
+    ner_dict = {"TP": [], "FP": [], "FN": []}
+    for tp in tps:
+        ner_dict["TP"].append({"Expected": tp[0], "Predicted": tp[1]})
```
Comment (erohmensing) [inline rasa/nlu/test.py]: Can we just add to the dict inside of the logic instead of creating lists and then adding the lists to the dict?

**Comment by tmbo** [comment]
@demongolem thanks a lot for the effort you already put into this, do you intend to take a look at the review comments to make sure we don't lose the time you put into this and get this merged?

**Comment by erohmensing** [comment]
Hi @demongolem, checking in. Are you still interested in contributing this PR?

**Comment by erohmensing** [comment]
closing due to inactivity and in favor of https://github.com/RasaHQ/rasa/issues/4329

**Pull request closed without merging**